### DISCUSSION TITLE
What is the lower amount of RAM rustic can be used with?

### DISCUSSION THREAD

**Discussion by parasiteoflife** [rustic-rs/rustic#1345] [Q&A] [answered]
I have a low power machine that I want to backup, the problem is that it doesn't have much RAM available, at most 400MB, so I'm wondering can I configure rustic in a way that it uses the less amount of RAM possible? And how much RAM will that be? I mean I know nobody can just see the future and tell me how much RAM rustic will use but I'm curious about the minimum RAM anyone got rustic working with.

I'm guessing I will have to disable compression? And use the default pack sizes if not smaller, but what else?

**Comment by NTmatter** [comment]
Would it be possible to get some info about your current dataset? Total/average size, file count, type of data, underlying file system(s)/raid?

Additionally, is swap space available? Any time constraints on backup runtime?

As a last resort, can you make your data available to another machine over the network? It might be possible to do the heavier lifting on another machine if the network is fast enough and dataset is small enough. Alternatively, an rsync/rclone to another host, running the backup out there.

**Reply by parasiteoflife** [reply]
The entire disk space is just 40GB, I estimate max 30GB will have to be back up, right now it must be at most 10GB. Can't say for sure how many files, FS is ext4, no RAID. Yes, I've enabled swap (4GB), backup can run all day if needed provided I can limit it to one core so other services aren't affected.

Does rustic works well with samba shares? Or should I use SSHFS? I thought it needed to access all files every time but I remember it checks for byte size and modtime which are available on SMB shares, I could use another machine to do the backup since the destination was going to be that same machine. The thing is what is rustic's performance with network shares, though I guess it couldn't be worst than using rustic on a RAM & CPU constrained environment

**Comment by aawsome** [comment, marked as answer]
@parasiteoflife While rustic is optimized to use as few memory as possible (and should have an overall memory footprint way lower than restic), I must say that the repository format in general is not optimized for very-low-memory-environments like yours. 

For many operations, the in-memory index is the main contributor to memory usage - this is largely memory-optimized in rustic, but it directly depends on the total repository size. So, if you have a large repository, you most likely will reach memory limits at some point...

For `backup`, rustic keeps approximately 4 pack files in memory before/while writing them, so reducing the pack size (e.g. using ing `rustic config --set-datapack-size-limit` or `--set-treepack-size-limit`) will help there. But, for large repositories, small pack sizes result in a (slightly) larger index size...

That said, I just checked the memory usage of a real-life repository having ~250GB repository size (with all default settings, i.e. a datapacksize of around 50MiB) and I am seeing memory usage below 100MB for standard operations like `rustic backup` or `rustic check`...

**Reply by parasiteoflife** [reply]
> That said, I just checked the memory usage of a real-life repository having ~250GB repository size (with all default settings, i.e. a datapacksize of around 50MiB) and I am seeing memory usage below 100MB for standard operations like rustic backup or rustic check...

Interesting, it seems rustic will at least backup without problems on this low amount of RAM available.

Could you please tell me what limitations or performance hit will rustic have if I mount the data as a samba/sshfs share? The machines are not on LAN so Internet will have to be used and the lowest end is at 400Mbps, I'm not too concerned on how much it will take at least for the first backup, but subsequent runs will perform close to local to local? Does rustic only checks for byte size and modtimes to detect modifications?

**Reply by aawsome** [reply]
Rustic checks for identical
- file type (file, dir, link, special file), obviously
- file size
- mtime
- ctime until using the `--ignore-ctime` option
- inode until using the `--ignore-inode` option

Unfortunately I have no clue about performance to get this information over a sambe/sshfs share..
